### ISSUE TITLE
Separate step "enter private key password" when key is encrypted with different password

### ISSUE THREAD

**Issue by let4be** [lavab/web#267]
Currently it's possible to change private key password: https://mail.lavaboom.co/#!/settings/security
under "Change password" section

When user has done so he will be able to login but his private key won't be decrypted automatically and he will have to do it manually on the settings tab.

I suggest to add a new screen to the Login application that would ask user for his private key password if private key's password differs from account's password.

What do you think @andreis, @flixi ?

**Comment by andreis** [comment]
We definitely need to work on the experience of decrypting/handling passphrases*.

On (1) login when LbSync active and (2) keyring upload, we should try to decrypt all available private keys, and warn the user if any private keys have different passphrases, then suggest to change the passphrase to the default one (the account password).

Also, when changing the account password, we should alert the user that the private keys passphrases will also be changed.

* I suggest we use **passphrase** to refer to the strings used to encrypt private keys

**Comment by let4be** [comment]
@andreis I thought it could be good to provide users a way to have private keys protected with passpharse different from account's password
Don't you think so?

**Comment by andreis** [comment]
Of course. We should make it easy to have private keys encrypted with the account password (the default use case), yet let power users customise their keys.

Additionally, when importing private keys (either by LS or uploading a keyring) we should indicate with an unread-emails-style icon next to settings how many keys we couldn't decrypt with the account password. Makes sense?

**Comment by let4be** [comment]
@andreis when we do import keys from http://mail.lavaboom.co/#!/settings/security
we don't have account password in plain text at all
(in some cases we could get access to it, for example when we came directly from the auth screen in the login application, but it's not guaranteed because we can authenticate via stored token)

So for now main application imports all private keys in encrypted state(from LS and backup) and user can manually enter password for each key

Additionally we trigger LS right after user logs in(at this stage we have password)

**Comment by andreis** [comment]
Side note, when are we going to get rid of the ugly `#!/` links already?

@let4be

**Comment by let4be** [comment]
@andreis pls...
create a separate issue instead going offtopic ^^

**Comment by alexouzounis** [comment]
tried to import private key with pass-phrase via http://mail.lavaboom.co/#!/settings/securityand i entered the pass-phrase in the relevant textbox when expanded the key ( after navigating out and in again the settings panel since no confirmation or reload of the page occurred ). Even then, the key icon is still red, as encrypted and when i logout and login again it is no longer there.

Am I missing something ? Many thanks for the help!

using OSX Yosemite, Firefox 38.0

**Comment by andreis** [comment]
Thank you for your bug report @alexouzounis 

Can you please check out https://mail.lavaboom.com/settings/security ? You seem to be reporting a bug for an old version of Lavaboom.

Also, screenshots and console logs are appreciated/rather necessary for diagnosis.

**Comment by alexouzounis** [comment]
* logging in to lavaboom ( https://mail.lavaboom.com/label/Inbox/ ), i see the following in the logs; http://pastebin.com/raw.php?i=jpWQBc7N ( errors about private key not found ) . 

* selecting settings --> Security, the console prints the following; http://pastebin.com/raw.php?i=UhmBpFaW ( looks clean to me )

* importing key via the import keys button, the console prints the following; http://pastebin.com/raw.php?i=S0AWC18m The UI does not refresh and the imported key is not visible, neither a confirmation that the key was imported successfully
![screen shot 2015-05-13 at 19 19 40](https://cloud.githubusercontent.com/assets/5610874/7617743/0d3524dc-f9a5-11e4-90bc-e2463ef4921e.png)

* selecting General and Security again and the key is now present; console: http://pastebin.com/raw.php?i=UaTPJu9r
![screen shot 2015-05-13 at 19 23 18](https://cloud.githubusercontent.com/assets/5610874/7617801/8c04abfc-f9a5-11e4-8fe3-f015a931cba2.png)

* I enter the passphrase ( same as the lavaboom pass for simplicity ), hit enter as there is not button to confirm and nothing happens; console: http://pastebin.com/raw.php?i=rB1dvsUZ
![screen shot 2015-05-13 at 19 27 43](https://cloud.githubusercontent.com/assets/5610874/7617897/2d7c7dca-f9a6-11e4-898a-e92e36579377.png)

* switching views between General and Settings and the pass-phrase is now gone; console: http://pastebin.com/raw.php?i=LkNt6p6R
![screen shot 2015-05-13 at 19 29 54](https://cloud.githubusercontent.com/assets/5610874/7617930/77a92e3e-f9a6-11e4-9a4e-6caefaae9287.png)

* logout & login ; console: http://pastebin.com/raw.php?i=qQi4NeZp

* navigating to Settings --> Security, the key is gone again; console: http://pastebin.com/raw.php?i=w2085fDf
![screen shot 2015-05-13 at 19 33 26](https://cloud.githubusercontent.com/assets/5610874/7617997/f3626cde-f9a6-11e4-96ca-f8dc29d20255.png)

Hopefully this will give you a bit more info :) Thanks again for looking into this.

**Comment by alexouzounis** [comment]
just realized that maybe lavaboom treats the first key as the private one so i repeated the process and what do you know; key decrypted message shows up after refreshing the page after the import as above and hitting enter when pass-phrase was entered. When I logout and login back again though the key is gone. Console: http://pastebin.com/raw.php?i=MwS84ZV7

* successful import
![screen shot 2015-05-13 at 19 49 42](https://cloud.githubusercontent.com/assets/5610874/7618358/993f5818-f9a9-11e4-9f7d-7535639cfa9f.png)

* logout and login, key is gone
![screen shot 2015-05-13 at 19 50 14](https://cloud.githubusercontent.com/assets/5610874/7618371/a8d538ce-f9a9-11e4-8a12-15b76cff3f14.png)

**Comment by andreis** [comment]
I'd suggest for now just using Lavaboom Sync, most of our testing is done with it enabled, and it works rather wlel. Deleting loaded keys on logout (if you don't tick the "private computer" box) is intended behaviour.

**Comment by andreis** [comment]
Anyway cc-ing @let4be on this

**Comment by let4be** [comment]
I"m going to check this in-depth tomorrow, but so far yes if Lavaboom Sync is enabled it now removes both encrypted and decrypted keys on page close(because both are being stored in session storage only)
+ we remove decrypted keys manually as soon as user logs out

One thing is, if sync is enabled it should automatically restore the synced key as soon as user logs in back...

**Comment by alexouzounis** [comment]
Cool, will try again with sync enabled. Where are the keys stored then ?

**Comment by andreis** [comment]
The keyring (encrypted private key + public key) is stored on our servers. The keys are encrypted with your password, so we can't access your keys without knowing your password.
